flashpay/apps/payments/views.py

```diff
@@ -131,3 +138,117 @@ def list(self, request: Request, *args: Any, **kwargs: Any) -> Response:
+        transactions = Transaction.objects.filter(
+            txn_ref=txn_ref, status=TransactionStatus.PENDING
+        )
+
+        if not transactions.exists():
+            return Response(
+                data={
+                    "status_code": status.HTTP_400_BAD_REQUEST,
+                    "message": "Transaction does not exist",  # noqa: E501
+                    "data": None,
+                },
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        transaction = transactions.first()
+
+        # check if transaction type of 'axfer' i.e transferring algo assets.
+        if api_response["transaction"]["tx-type"] != "axfer":
+            return Response(
+                data={
+                    "status_code": status.HTTP_400_BAD_REQUEST,
+                    "message": "An error occured while setting up your account due to bad transaction provided.",  # noqa: E501
+                    "data": None,
+                },
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        # check if the sender of the txn is the same as the one provided.
+        if api_response["transaction"]["sender"] != transaction.sender:  # type: ignore
+            return Response(
+                data={
+                    "status_code": status.HTTP_400_BAD_REQUEST,
+                    "message": "An error occured while setting up your account due to bad transaction provided.",  # noqa: E501
+                    "data": None,
+                },
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        # check if the receiver of the txn is same as the one provided.
+        if (
+            api_response["transaction"]["asset-transfer-transaction"]["receiver"]
+            != transaction.recipient  # type: ignore
+        ):
+            return Response(
+                data={
+                    "status_code": status.HTTP_400_BAD_REQUEST,
+                    "message": "An error occured while setting up your account due to bad transaction provided.",  # noqa: E501
+                    "data": None,
+                },
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        # check if the txn asset and amount is same as the one provided.
+        if (
+            api_response["transaction"]["asset-transfer-transaction"]["amount"]
+            != (transaction.amount * 100)  # type: ignore
+        ) or (
+            api_response["transaction"]["asset-transfer-transaction"]["asset-id"]
+            != (transaction.asset.asa_id)  # type: ignore
+        ):
+            return Response(
+                data={
+                    "status_code": status.HTTP_400_BAD_REQUEST,
+                    "message": "An error occured while setting up your account due to bad transaction provided.",  # noqa: E501
+                    "data": None,
+                },
+                status=status.HTTP_400_BAD_REQUEST,
+            )
```

Comment:
This could raise a potential bug in the code.
1. What if `asset-transfer-transaction` is not in the `transaction` field? (is that not a possibility ?)
2. And we might need to know the decimals of the `Assets` to know what to multiply by. We can easily add that to the `Assets` table or if the info is available on the chain and in the `api_response` we can use it. Different assets have different decimals I guess.

Reply:
it should be on the Assets table then

Reply:
1. The preceding block of code checks if it is an `asset-transfer-transaction`

Reply:
Precisely... you can add that in your changes

Reply:
> 1. The preceding block of code checks if it is an `asset-transfer-transaction`

But it's not nested in it. Meaning if it's not an asset transfer, it's most likely to fail.. you can test the codes with variable conditions in your test cases so you can be sure it works

Reply:
> > 1. The preceding block of code checks if it is an `asset-transfer-transaction`
> 
> But it's not nested in it. Meaning if it's not an asset transfer, it's most likely to fail.. you can test the codes with variable conditions in your test cases so you can be sure it works

only two transaction types should be allowed, asset-transfer `axfer` and algo transfer `pay`.